iroha/src/sumeragi.rs

```diff
@@ -656,343 +659,561 @@ impl<Q: QueueTrait, G: GenesisNetworkTrait, W: WorldTrait> Sumeragi<Q, G, W> {
+
+        pub fn build(self) -> Result<Topology> {
+            let peers = field_is_some_or_err!(self.peers)?;
+            let max_faults = field_is_some_or_err!(self.max_faults)?;
+            let reshuffle_after_n_view_changes =
+                field_is_some_or_err!(self.reshuffle_after_n_view_changes)?;
+            let at_block = field_is_some_or_err!(self.at_block)?;
+
+            let min_peers = 3 * max_faults + 1;
+            let peers: Vec<_> = peers.into_iter().collect();
+            if peers.len() >= min_peers as usize {
+                let sorted_peers = if self.n_view_changes > reshuffle_after_n_view_changes {
+                    sort_peers_by_hash_and_counter(peers, at_block, self.n_view_changes)
+                } else {
+                    let peers = sort_peers_by_hash(peers, at_block);
+                    shift_peers_by_n(peers, self.n_view_changes)
+                };
+                Ok(Topology {
+                    sorted_peers,
+                    max_faults,
+                    reshuffle_after_n_view_changes,
+                    at_block,
+                    n_view_changes: self.n_view_changes,
+                })
+            } else {
+                Err(error!(
+                "Not enough peers to be Byzantine fault tolerant. Expected a least {} peers, got {}",
+                min_peers,
+                peers.len(),
+            ))
+            }
         }
-        let Hash(bytes) = Hash::new(&bytes);
-        let mut rng = StdRng::from_seed(bytes);
-        self.sorted_peers.shuffle(&mut rng);
     }
 
-    /// Shifts `sorted_peers` by one to the right.
-    #[allow(clippy::expect_used)]
-    pub fn shift_peers_by_one(&mut self) {
-        let last_element = self
-            .sorted_peers
-            .pop()
-            .expect("No elements found in sorted peers.");
-        self.sorted_peers.insert(0, last_element);
+    #[derive(Clone, Debug, Encode, Decode)]
+    pub struct Topology {
+        /// Current order of peers. The roles of peers are defined based on this order.
+        sorted_peers: Vec<PeerId>,
+        /// Maximum faulty peers in a network.
+        max_faults: u32,
+
+        reshuffle_after_n_view_changes: u32,
+
+        at_block: Hash,
+
+        n_view_changes: u32,
     }
 
-    /// Shifts `sorted_peers` by `n` to the right.
-    pub fn shift_peers_by_n(&mut self, n: u32) {
-        for _ in 0..n {
-            self.shift_peers_by_one();
+    impl Topology {
+        pub fn builder() -> Builder {
+            Builder::new()
         }
-    }
 
-    /// Get role of the peer by its id.
-    pub fn role(&self, peer_id: &PeerId) -> Role {
-        if self.leader() == peer_id {
-            Role::Leader
-        } else if self.proxy_tail() == peer_id {
-            Role::ProxyTail
-        } else if self.validating_peers().contains(peer_id) {
-            Role::ValidatingPeer
-        } else {
-            Role::ObservingPeer
+        pub fn into_builder(self) -> Builder {
```

Comment:
Kinda don't like this method. Builder shouldn't be used for mutation of topology. It should be used to build new one.